<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Import-Raw-Data" data-toc-modified-id="Import-Raw-Data-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Import Raw Data</a></span><ul class="toc-item"><li><span><a href="#Raw-Data-Describtion" data-toc-modified-id="Raw-Data-Describtion-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Raw Data Describtion</a></span></li></ul></li><li><span><a href="#Data-Preprocessing" data-toc-modified-id="Data-Preprocessing-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data Preprocessing</a></span><ul class="toc-item"><li><span><a href="#Dealing-with-some-non-numeric-features" data-toc-modified-id="Dealing-with-some-non-numeric-features-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Dealing with some non-numeric features</a></span><ul class="toc-item"><li><span><a href="#Baseline-Dataset" data-toc-modified-id="Baseline-Dataset-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Baseline Dataset</a></span><ul class="toc-item"><li><span><a href="#TOBACCO_STATUS_LABEL" data-toc-modified-id="TOBACCO_STATUS_LABEL-2.1.1.1"><span class="toc-item-num">2.1.1.1&nbsp;&nbsp;</span>TOBACCO_STATUS_LABEL</a></span></li><li><span><a href="#RACE" data-toc-modified-id="RACE-2.1.1.2"><span class="toc-item-num">2.1.1.2&nbsp;&nbsp;</span>RACE</a></span></li><li><span><a href="#BLOOD_TYPE" data-toc-modified-id="BLOOD_TYPE-2.1.1.3"><span class="toc-item-num">2.1.1.3&nbsp;&nbsp;</span>BLOOD_TYPE</a></span></li><li><span><a href="#FINANCIAL_CLASS_CODE" data-toc-modified-id="FINANCIAL_CLASS_CODE-2.1.1.4"><span class="toc-item-num">2.1.1.4&nbsp;&nbsp;</span>FINANCIAL_CLASS_CODE</a></span></li></ul></li><li><span><a href="#Admissions-Dataset" data-toc-modified-id="Admissions-Dataset-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>Admissions Dataset</a></span><ul class="toc-item"><li><span><a href="#PATIENT_TYPE_CODE" data-toc-modified-id="PATIENT_TYPE_CODE-2.1.2.1"><span class="toc-item-num">2.1.2.1&nbsp;&nbsp;</span>PATIENT_TYPE_CODE</a></span></li><li><span><a href="#HOSP_ROLLUP_GROUPING1" data-toc-modified-id="HOSP_ROLLUP_GROUPING1-2.1.2.2"><span class="toc-item-num">2.1.2.2&nbsp;&nbsp;</span>HOSP_ROLLUP_GROUPING1</a></span></li></ul></li><li><span><a href="#Drop-all-non-numeric-features" data-toc-modified-id="Drop-all-non-numeric-features-2.1.3"><span class="toc-item-num">2.1.3&nbsp;&nbsp;</span>Drop all non-numeric features</a></span></li></ul></li><li><span><a href="#Dealing-with-Missing-Values" data-toc-modified-id="Dealing-with-Missing-Values-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Dealing with Missing Values</a></span><ul class="toc-item"><li><span><a href="#Drop-columns-with-more-than-1%-NA-values" data-toc-modified-id="Drop-columns-with-more-than-1%-NA-values-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Drop columns with more than 1% NA values</a></span></li><li><span><a href="#Filling-'HOSP_CR',-'HOSP_NA',-'HOSP_EGFR'-with-mean-values" data-toc-modified-id="Filling-'HOSP_CR',-'HOSP_NA',-'HOSP_EGFR'-with-mean-values-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Filling 'HOSP_CR', 'HOSP_NA', 'HOSP_EGFR' with mean values</a></span></li><li><span><a href="#Filling-'ASA_00',-'INSULIN_00-',-...-with-0" data-toc-modified-id="Filling-'ASA_00',-'INSULIN_00-',-...-with-0-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>Filling 'ASA_00', 'INSULIN_00 ', ... with 0</a></span></li></ul></li><li><span><a href="#Join-2-datasets-together" data-toc-modified-id="Join-2-datasets-together-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Join 2 datasets together</a></span><ul class="toc-item"><li><span><a href="#Deal-with-procudures-during-hospitalization-features" data-toc-modified-id="Deal-with-procudures-during-hospitalization-features-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Deal with procudures during hospitalization features</a></span></li><li><span><a href="#Concatenate-two-datasets-together" data-toc-modified-id="Concatenate-two-datasets-together-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Concatenate two datasets together</a></span></li></ul></li><li><span><a href="#Normalize-data" data-toc-modified-id="Normalize-data-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Normalize data</a></span></li></ul></li><li><span><a href="#Logistic-Regression" data-toc-modified-id="Logistic-Regression-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Logistic Regression</a></span></li><li><span><a href="#Random-Forest" data-toc-modified-id="Random-Forest-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Random Forest</a></span></li><li><span><a href="#Feedforward-Neural-Net" data-toc-modified-id="Feedforward-Neural-Net-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Feedforward Neural Net</a></span></li><li><span><a href="#Voting-Classifier" data-toc-modified-id="Voting-Classifier-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Voting Classifier</a></span></li><li><span><a href="#SVM" data-toc-modified-id="SVM-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>SVM</a></span></li><li><span><a href="#Visualization" data-toc-modified-id="Visualization-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Visualization</a></span></li></ul></div>

In [8]:
import pandas as pd
import numpy as np
import sklearn

In [9]:
pd.options.display.max_rows = 400
pd.options.display.max_columns = 50

# Import Raw Data
- raw_baselinedt
- raw_admissiondt

HF-Baseline:\
Represents the initial encounter with UPMC with a diagnosis of hear failure

HF-Admissions: \
Represents subsequent admissions of those patients with heart failure \

12285 patients

In [10]:
%cd datasets

[Errno 2] No such file or directory: 'datasets'
/Users/jinchenxie/Desktop/Research/datasets


In [11]:
raw_baselinedt = pd.read_excel('HF-Baseline.xlsx', '00480_Baseline_20200115_NA')

raw_admissiondt = pd.read_excel('HF-Admissions.xlsx', '00480_Admissions_20200115_NA')

## Raw Data Describtion

In [12]:
print("the baseline dataset has shape", raw_baselinedt.shape)
print("the admissions dataset has shape", raw_admissiondt.shape)

the baseline dataset has shape (12285, 320)
the admissions dataset has shape (55742, 168)


In [13]:
raw_baselinedt.head()

rID  LOS_DAYS  HOSP_CR  HOSP_BILI  HOSP_INR  HOSP_NA  HOSP_EGFR  \
0    1     6.198     3.34        NaN       NaN    136.0      17.75   
1    2     3.240     1.40        2.8       1.4    136.0      47.45   
2    3     4.074     2.76        2.1       3.0    141.0      22.46   
3    4     8.128     1.89        NaN       7.6    139.0      35.85   
4    5     3.121     1.04        1.0       1.1    141.0      91.09   

  PRIMARY_DIAGNOSIS_CODE PRIMARY_DIAGNOSIS_DESCRIPTION  AGE_ADMISSION  \
0                 I50.43     ACUTE ON CHRONIC COMBINED             82   
1                 428.21       AC SYSTOL HEART FAILURE             91   
2                 428.33       AC-CHR DIASTOL HRT FAIL             76   
3                 I50.33    ACUTE ON CHRONIC DIASTOLIC             66   
4                 I50.23     ACUTE ON CHRONIC SYSTOLIC             51   

   AGE_OFFICE_VISIT  FEMALE   RACE            ETHNIC_GROUP    BMI  WEIGHT_KG  \
0              82.9       0  White  Not Hispanic or Latino  28.28      84.37   
1              91.6       0  White  Not Hispanic or Latino  21.41      60.65   
2              76.9       0  White  Not Hispanic or Latino  28.58      80.29   
3              66.1       0  White  Not Hispanic or Latino  31.32      90.72   
4              51.3       0  Black  Not Hispanic or Latino  44.74     153.77   

   ORD  DAYS_TO_CLOSEST_OFFICE_VISIT      DEPARTMENT_NAME  \
0    1                            14  XMEDICOR ASSOC ERIE   
1    1                            35          GM HBC SHEA   
2    1                            16          SFM HAMPTON   
3    1                             9             CA HAMOT   
4    1                           -63       XHVI MERCY OFC   

           EXTERNAL_DEPARTMENT_NAME DEPARTMENT_SPECIALTY  BP_SYSTOLIC  \
0           XMedicor Associates Inc           Cardiology        118.0   
1               Shea Medical Center    Internal Medicine        110.0   
2           SUPERIOR FAM MD HAMPTON      Family Practice        100.0   
3  UPMC Hamot Flagship CVT Surgeons      Cardiac Surgery        108.0   
4                      HVI at Mercy           Cardiology        140.0   

   BP_DIASTOLIC  PULSE FINANCIAL_CLASS_CODE  ... BNP_DAYS BILI BILI_DAYS  INR  \
0          58.0   72.0                    M  ...    -14.0  NaN       NaN  NaN   
1          58.0   84.0                    M  ...      NaN  0.7       1.0  1.1   
2          50.0   80.0                   MS  ...    -16.0  2.1     -15.0  1.3   
3          62.0   90.0                    M  ...     -9.0  0.9     -23.0  2.0   
4         100.0   78.0                   UX  ...    -19.0  NaN       NaN  NaN   

   INR_DAYS  BLOOD_TYPE  CLOSEST_MELD  VERIFIED_MELD  MELD_EST_MORTALITY  \
0       NaN         AB+           NaN            NaN                 NaN   
1     -11.0         NaN     18.059244           -1.0                 6.0   
2       7.0         NaN     31.254738           -1.0                52.6   
3      -1.0          O+     23.323969            0.0                19.6   
4       NaN         NaN      7.872816           -1.0                 1.9   

   RHC_FOUND  RA_MEAN_PRESSURE  RA_MEAN_PRESSURE_DAYS  THERMAL_CARDIAC_INDEX  \
0          0               NaN                    NaN                    NaN   
1          0               NaN                    NaN                    NaN   
2          0               NaN                    NaN                    NaN   
3          0               NaN                    NaN                    NaN   
4          0               NaN                    NaN                    NaN   

   THERMAL_CARDIAC_INDEX_DAYS  FICK_CARDIAC_INDEX  FICK_CARDIAC_INDEX_DAYS  \
0                         NaN                 NaN                      NaN   
1                         NaN                 NaN                      NaN   
2                         NaN                 NaN                      NaN   
3                         NaN                 NaN                      NaN   
4                         NaN            

- Baseline dataset has 320 features including 'rID'. 291 numerical features, and 29 string features.
- Admissions dataset has 168 features also including 'rID'. 143 numerical features, and 25 string features.
- Each patient has exactly 1 row in baseline dataset, but can have multiple rows in admissions. 

# Data Preprocessing

## Dealing with some non-numeric features

### Baseline Dataset

In [14]:
obj_bs = raw_baselinedt.select_dtypes(include=['object']).copy()
obj_bs.head()

PRIMARY_DIAGNOSIS_CODE PRIMARY_DIAGNOSIS_DESCRIPTION   RACE  \
0                 I50.43     ACUTE ON CHRONIC COMBINED  White   
1                 428.21       AC SYSTOL HEART FAILURE  White   
2                 428.33       AC-CHR DIASTOL HRT FAIL  White   
3                 I50.33    ACUTE ON CHRONIC DIASTOLIC  White   
4                 I50.23     ACUTE ON CHRONIC SYSTOLIC  Black   

             ETHNIC_GROUP      DEPARTMENT_NAME  \
0  Not Hispanic or Latino  XMEDICOR ASSOC ERIE   
1  Not Hispanic or Latino          GM HBC SHEA   
2  Not Hispanic or Latino          SFM HAMPTON   
3  Not Hispanic or Latino             CA HAMOT   
4  Not Hispanic or Latino       XHVI MERCY OFC   

           EXTERNAL_DEPARTMENT_NAME DEPARTMENT_SPECIALTY FINANCIAL_CLASS_CODE  \
0           XMedicor Associates Inc           Cardiology                    M   
1               Shea Medical Center    Internal Medicine                    M   
2           SUPERIOR FAM MD HAMPTON      Family Practice                   MS   
3  UPMC Hamot Flagship CVT Surgeons      Cardiac Surgery                    M   
4                      HVI at Mercy           Cardiology                   UX   

       FINANCIAL_CLASS HOSP_ROLLUP_GROUPING1 HOSP_ROLLUP_GROUPING2  \
0      MEDICARE PART A              Medicare              Medicare   
1      MEDICARE PART A              Medicare              Medicare   
2    SECURITY BLUE HMO         Security Blue              Medicare   
3      MEDICARE PART A              Medicare              Medicare   
4  UPMC EX SEL REG HMO       UPMC Commercial            Commercial   

  CLOSEST_PS_DAYS CLOSEST_TS_DAYS PRE_TRANSPLANT_DAYS PRIOR_PCI_DAYS  \
0             NaN             NaN                 NaN            NaN   
1             NaN             NaN                 NaN            NaN   
2             NaN             NaN                 NaN            NaN   
3             NaN             NaN                 NaN            NaN   
4             NaN             NaN                 NaN            NaN   

  PRIOR_CABG_DAYS PRIOR_TAVR_DAYS PRIOR_SAVR_MECH_DAYS  \
0             NaN             NaN                  NaN   
1             NaN             NaN                  NaN   
2             NaN             NaN                  NaN   
3             NaN             NaN                  NaN   
4             NaN             NaN                  NaN   

  PRIOR_SURG_AO_GRFT_DAYS PRIOR_ENDO_AO_GRFT_DAYS PRIOR_AO_VLV_REPAIR_DAYS  \
0                     NaN                     NaN                      NaN   
1                     NaN                     NaN                      NaN   
2                     NaN                     NaN                      NaN   
3                     NaN                     NaN                      NaN   
4                     NaN                     NaN                      NaN   

  PRIOR_SMVR_DAYS PRIOR_SPVR_DAYS PRIOR_STVR_DAYS      TOBACCO_STATUS_LABEL  \
0             NaN             NaN             NaN  QUIT (UNKNOWN TIMEFRAME)   
1             NaN             NaN             NaN                     NEVER   
2             NaN             NaN             NaN                     NEVER   
3               -             NaN             NaN  QUIT (UNKNOWN TIMEFRAME)   
4             NaN             NaN             NaN                   CURRENT   

  HF_ETIOLOGY_LABEL LDL_DAYS PREALBUMIN_DAYS BLOOD_TYPE  
0             OTHER      NaN             NaN        AB+  
1             OTHER      NaN             NaN        NaN  
2             OTHER      NaN             NaN        NaN  
3             OTHER      NaN               -         O+  
4             OTHER      NaN             NaN        NaN

Possible Categorical features to encode:
- RACE
- FINANCIAL_CLASS_CODE
- BLOOD_TYPE
- TOBACCO_STATUS_LABEL

#### TOBACCO_STATUS_LABEL

In [15]:
obj_bs['TOBACCO_STATUS_LABEL'].value_counts()#### BLOOD_TYPE

QUIT (UNKNOWN TIMEFRAME)        5216
NEVER                           4579
CURRENT                         1181
REMOTE (MORE THAN 12 MONTHS)     502
RECENT (WITHIN 12 MONTHS)        425
UNKNOWN                          266
NEVER ASKED                       82
PASSIVE                           34
Name: TOBACCO_STATUS_LABEL, dtype: int64

In [16]:
# Level encoding
clean_nums = {"TOBACCO_STATUS_LABEL": {"NEVER": 0, "QUIT (UNKNOWN TIMEFRAME)": 1,
                                   "REMOTE (MORE THAN 12 MONTHS)": 1, "RECENT (WITHIN 12 MONTHS)": 1,
                                  "UNKNOWN": 2, "NEVER ASKED": 2,
                                  "PASSIVE": 2, "CURRENT": 3}}
obj_bs.replace(clean_nums, inplace=True)
obj_bs['TOBACCO_STATUS_LABEL'].value_counts()

1    6143
0    4579
3    1181
2     382
Name: TOBACCO_STATUS_LABEL, dtype: int64

#### RACE

In [17]:
obj_bs.RACE.isna().sum()
obj_bs['RACE'].value_counts()

White      10682
Black       1466
Decline       32
Not Spe       28
Other A       16
Indian         9
America        5
Vietnam        4
Filipin        4
Chinese        4
Alaska         3
Korean         1
Other P        1
Name: RACE, dtype: int64

In [18]:
# Race categories encoding
obj_bs = obj_bs.fillna({"RACE": "Other P"})
clean_nums = {"RACE": {"Others": 0,"Decline": 0,"Not Spe": 0,"Other A": 0,"Indian": 0,
                       "America": 0,"Vietnam": 0,"Filipin": 0,"Chinese": 0,"Alaska": 0,
                       "Other P": 0,"Korean": 0, "Black": 0,
                                   "White": 1}}
obj_bs.replace(clean_nums, inplace=True)
obj_bs['RACE'].value_counts()

1    10682
0     1603
Name: RACE, dtype: int64

#### BLOOD_TYPE

In [19]:
obj_bs.BLOOD_TYPE.isna().sum()

8619

#### FINANCIAL_CLASS_CODE

In [20]:
obj_bs['FINANCIAL_CLASS_CODE'].value_counts()

M     4144
MS    2469
UM    1966
Q      483
AD     472
BE     335
UH     310
GG     230
2      189
MC     187
MU     136
S      128
GA     124
US      83
XL      81
P       81
2M      74
UY      63
2S      58
UE      56
0       55
HA      49
UN      42
4H      39
CG      37
MP      30
C       28
B       28
UX      27
3B      24
VA      20
UT      20
CB      16
4N      14
SA      13
AT      12
2G      11
2A      11
6       10
2C       9
AP       9
3G       7
SM       7
H        7
R        6
3L       5
AC       5
2U       4
4U       4
D        4
HP       4
N        4
CM       4
4A       3
1        3
7        3
IG       3
HH       3
4C       3
4        3
2B       2
4X       2
GM       2
3        2
6P       1
HU       1
W        1
3U       1
1B       1
5        1
BM       1
4R       1
3H       1
K        1
2N       1
T        1
3P       1
UC       1
V        1
BS       1
4M       1
G        1
2E       1
3M       1
OR       1
HG       1
Name: FINANCIAL_CLASS_CODE, dtype: int64

    We turned 'RACE' and 'TOBACCO_STATUS_LABEL' to numeric categories for BASELINE dataset.
   

### Admissions Dataset

In [21]:
obj_ad = raw_admissiondt.select_dtypes(include=['object']).copy()

#### PATIENT_TYPE_CODE

In [22]:
obj_ad['PATIENT_TYPE_CODE'].value_counts()


I     54806
S       873
E        53
DO       10
Name: PATIENT_TYPE_CODE, dtype: int64

In [23]:
clean_nums = {"PATIENT_TYPE_CODE": {"I": 1, "S": 0, "E": 0, "DO": 0}}
obj_ad.replace(clean_nums, inplace=True)
obj_ad['PATIENT_TYPE_CODE'].value_counts()

1    54806
0      936
Name: PATIENT_TYPE_CODE, dtype: int64

#### HOSP_ROLLUP_GROUPING1

In [24]:
obj_ad['HOSP_ROLLUP_GROUPING1'].value_counts()

Medicare              21728
UPMC MC Managed       10136
Security Blue          9311
Other MC Managed       5797
UPMC MA Managed        2008
Highmark Commercia     1860
UPMC Commercial        1647
National Payers         845
Gateway                 766
Medicaid                707
Other MA Managed        306
Other                   258
Self Pay                188
Other Commercial        183
MA Pending                2
Name: HOSP_ROLLUP_GROUPING1, dtype: int64

    We turned 'PATIENT_TYPE_CODE' into numeric categories.

### Drop all non-numeric features

In [25]:
baseline_num = raw_baselinedt.drop(list(raw_baselinedt.select_dtypes('object').columns), axis = 1) 
obj_bs = obj_bs.drop(list(obj_bs.select_dtypes('object').columns), axis = 1) 
baseline_num = pd.concat([baseline_num, obj_bs], axis=1, join='inner')

In [26]:
admissions_num = raw_admissiondt.drop(list(raw_admissiondt.select_dtypes('object').columns), axis = 1)
obj_ad = obj_ad.drop(list(obj_ad.select_dtypes('object').columns), axis = 1)
admissions_num = pd.concat([admissions_num, obj_ad], axis=1, join='inner')

## Dealing with Missing Values

In [27]:
print("the all-numeric baseline dataset has shape", baseline_num.shape)
print("the all-numeric admissions dataset has shape", admissions_num.shape)

the all-numeric baseline dataset has shape (12285, 293)
the all-numeric admissions dataset has shape (55742, 144)


### Drop columns with more than 1% NA values

In [28]:
baseline_clean = baseline_num.dropna(thresh=baseline_num.shape[0]-122, axis=1)
admissions_clean = admissions_num.dropna(thresh=admissions_num.shape[0]-122, axis=1)

In [29]:
baseline_clean.shape

(12285, 131)

### Filling 'HOSP_CR', 'HOSP_NA', 'HOSP_EGFR' with mean values

In [30]:
baseline_clean['HOSP_CR'].fillna(baseline_clean['HOSP_CR'].mean(), inplace=True)
baseline_clean['HOSP_EGFR'].fillna(baseline_clean['HOSP_EGFR'].mean(), inplace=True)
baseline_clean['HOSP_NA'].fillna(baseline_clean['HOSP_NA'].mean(), inplace=True)

/Users/jinchenxie/miniconda3/envs/research_env/lib/python3.6/site-packages/pandas/core/generic.py:6287: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


### Filling 'ASA_00', 'INSULIN_00 ', ... with 0

In [31]:
baseline_clean = baseline_clean.fillna(0)

In [32]:
print("After dropping and filling NA values:")
print("Baseline and admissions datasets have shape ",baseline_clean.shape, 
      'and' ,admissions_clean.shape, 'respectively.')

After dropping and filling NA values:
Baseline and admissions datasets have shape  (12285, 131) and (55742, 43) respectively.


## Join 2 datasets together

### Deal with procudures during hospitalization features

- Some are included in both baseline and admissions datasets:\
CBGG, TAVR, SAVR_MECH, ... CRT_IMPLANT 
<br>

- Some are only indluded in admissions datasets:\
PCI, IMPELLA. IABP, ..., BIV_ICD_IMPLANT

For procedures that appeared both in admissions and baseline datasets, we want to set the feature=1 if the feature in either dataset contains 1. \
e.g. IF_CABG = MAX( MAX(admissions.CABG), baseline.PRIOR_CABG)

In [33]:
# This function used to deal with procedure features that appeared in both baseline and admissions
def if_procedure(proc, prior_proc, new_feature, baseline, admissions):
    """
    proc: name of the procudure record in admission data
    prior_proc: name of the procedure record in baseline data
    new_feature: name of the new feature
    baseline, admissions: baseline and adimissions datasets
    @return: altered baseline dataset
    """
    adm_patients = np.asarray(admissions.loc[admissions[proc]==1].rID)
    baseline[new_feature] = 0
    for i in np.unique(adm_patients):
        baseline.at[i-1, new_feature] = 1
    return baseline

In [34]:
baseline_v1 = baseline_clean.copy()

In [35]:
list_proc = ['CABG','TAVR','SAVR_MECH','SAVR_BIO','SURG_AO_GRFT','ENDO_AO_GRFT','AO_VLV_REPAIR',
            'SMVR','SPVR','STVR','POST_VAD','POST_TRANSPLANT','CRT_IMPLANT']
list_preproc = ['PRIOR_CABG','PRIOR_TAVR','PRIOR_SAVR_MECH','PRIOR_SAVR_BIO','PRIOR_SURG_AO_GRFT',
               'PRIOR_ENDO_AO_GRFT','PRIOR_AO_VLV_REPAIR','PRIOR_SMVR','PRIOR_SPVR',
               'PRIOR_STVR','PRE_VAD','PRE_TRANSPLANT','PRE_CRT']
list_new_feature = ['IF_CABG','IF_TAVR','IF_SAVTMECH','IF_SAVRBIO','IF_SURG_AOGRFT','IF_ENDO_AOGRFT',
                   'IF_AOVLV_REPAIR','IF_SMVR','IF_SPVR','IF_STVR','IF_POSTVAD',
                   'IF_POTRANSPLANT','IF_CRTIMPLANT']

In [36]:
for i in range(len(list_proc)):
    print('doing',list_proc[i])
    baseline_v1 = if_procedure(list_proc[i],list_preproc[i],list_new_feature[i],
                              baseline_v1, admissions_clean)
baseline_v1.drop(list_preproc, axis=1, inplace=True)

doing CABG
doing TAVR
doing SAVR_MECH
doing SAVR_BIO
doing SURG_AO_GRFT
doing ENDO_AO_GRFT
doing AO_VLV_REPAIR
doing SMVR
doing SPVR
doing STVR
doing POST_VAD
doing POST_TRANSPLANT
doing CRT_IMPLANT


For procedures that only appeared in admissions dataset, we add new features in baseline_v1.\
baseline_v1.new_feature = 1 if MAX(admissions.feature == 1)

In [37]:
baseline_v1.shape

(12285, 131)

In [38]:
def if_procedure2(proc, new_feature, baseline, admissions):
    """
    proc: feature name of the procedure in admissions dataset
    new_feature: the name of the procedure add into baseline_v1
    baseline, admissions: baseline and admissions datasets
    """
    adm_patients = np.asarray(admissions.loc[admissions[proc]==1].rID)
    baseline[new_feature] = 0
    for i in np.unique(adm_patients):
#         print(i)
        baseline.at[i-1, new_feature] = 1
    return baseline

In [39]:
list_proc2 = ['PCI','IMPELLA','IABP','ECMO','ABLATION','PACEMAKER_IMPLANT',
              'ICD_IMPLANT','BIV_ICD_IMPLANT']
list_new_features2 = ['IF_PCI','IF_IMPELLA','IF_IABP','IF_ECMO','IF_ABLATION',
                     'IF_PACEMAKER','IF_ICDIMPLANT','IF_BIV_ICDIMPLANT']

In [40]:
for i in range(len(list_proc2)):
    print('doing',list_proc2[i])
    baseline_v1 = if_procedure2(list_proc2[i], list_new_features2[i],
                               baseline_v1, admissions_clean)

doing PCI
doing IMPELLA
doing IABP
doing ECMO
doing ABLATION
doing PACEMAKER_IMPLANT
doing ICD_IMPLANT
doing BIV_ICD_IMPLANT


In [41]:
admission_v1 = admissions_clean.copy()

admission_v1.drop(list_proc, axis=1, inplace=True)
admission_v1.drop(list_proc2, axis=1, inplace=True)

admission_v1.drop('INDEX_HOSPITALIZATION', axis=1, inplace=True)

In [42]:
admission_v1.shape

(55742, 21)

In [43]:
# Admission_v2 group by patient ID and take the max of each feature
admission_v1 = admission_v1.groupby('rID').max()

### Concatenate two datasets together

In [44]:
dataset_v1 = pd.merge(left=baseline_v1, right=admission_v1, left_on='rID', right_on='rID')

In [55]:
dataset_v1.drop('AGE_ADMISSION_y', axis=1,inplace=True)

KeyError: "['AGE_ADMISSION_y'] not found in axis"

In [56]:
dataset_v1.shape

(12285, 158)

In [47]:
from sklearn import preprocessing

In [48]:
import sklearn.model_selection as model_selection
from sklearn.linear_model import LogisticRegressionCV

from sklearn.metrics import accuracy_score

In [62]:
dataset_v2 = dataset_v1.copy()
for i in range(dataset_v2.shape[1]):
    if (dataset_v2.iloc[:,i].max()==1):
        pass
    else:
        dataset_v2.iloc[:,i] = (dataset_v2.iloc[:,i]-dataset_v2.iloc[:,i].mean())/dataset_v2.iloc[:,i].std()

In [63]:
dataset_v2.to_excel('clean_HFdata_v1.xlsx', index = False, header = True, 
                    sheet_name='sheet1')

## Normalize data

In [37]:
# dataset_norm = dataset_v1.copy()
# for i in range(dataset_norm.shape[1]):
#     if (dataset_norm.iloc[:,i].max()==1):
#         pass
#     else:
#         dataset_norm.iloc[:,i] = (dataset_norm.iloc[:,i]-dataset_norm.iloc[:,i].mean())/dataset_norm.iloc[:,i].std()

In [38]:
# Set mortality status as target
y = dataset_v1['CTB']
X = dataset_v1.drop('CTB', 1)

In [39]:
# Remove the patient ID label
X = X.drop('rID',1)

In [40]:
X = X.replace([np.inf, -np.inf], np.nan)
X.fillna(8.475169e-01, inplace=True)

# Logistic Regression

In [45]:
X_scaled = preprocessing.scale(X)

In [62]:
X_train, X_test, y_train, y_test = model_selection.train_test_split(X_scaled, y, 
                                        train_size=0.80, test_size=0.20, random_state=101)

In [47]:
model_lr = LogisticRegressionCV(10, max_iter=900)
model_lr.fit(X_train, y_train)

/Users/jinchenxie/miniconda3/envs/research_env/lib/python3.6/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)


LogisticRegressionCV(Cs=10, class_weight=None, cv='warn', dual=False,
                     fit_intercept=True, intercept_scaling=1.0, l1_ratios=None,
                     max_iter=900, multi_class='warn', n_jobs=None,
                     penalty='l2', random_state=None, refit=True, scoring=None,
                     solver='lbfgs', tol=0.0001, verbose=0)

In [48]:
LR_predicted = model_lr.predict(X_test)
LR_accuracy = accuracy_score(y_test, LR_predicted)

In [49]:
print('the Logistic Regression prediction has accuracy', LR_accuracy)

the Logistic Regression prediction has accuracy 0.8595848595848596


# Random Forest

In [50]:
from sklearn.ensemble import RandomForestClassifier

In [51]:
model_rf = RandomForestClassifier(n_estimators = 50, random_state = 0, max_depth=30)

In [52]:
model_rf.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=30, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=50,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [53]:
RF_predicted = model_rf.predict(X_test)
RF_accuracy = accuracy_score(y_test, RF_predicted)

In [54]:
print('the Random Forest Classifier has accuracy', RF_accuracy)

the Random Forest Classifier has accuracy 0.8616198616198616


# Feedforward Neural Net

In [59]:
import torch
from torch.utils.data import Dataset, DataLoader, TensorDataset
import torch.nn as nn
import torch.nn.functional as F

In [68]:
X_train = pd.DataFrame(X_train)
X_test = pd.DataFrame(X_test)
y_train = pd.DataFrame(y_train)
y_test = pd.DataFrame(y_test)

In [69]:
# Prepare the data for torch nn
train_x = torch.tensor(X_train.values.astype(np.float32))
train_y = torch.tensor(y_train.values.astype(np.int_))

test_x = torch.tensor(X_test.values.astype(np.float32))
test_y = torch.tensor(y_test.values.astype(np.int_))

train_data = TensorDataset(train_x, train_y)
train_loader = DataLoader(dataset=train_data, batch_size=36, shuffle=True)

validation_data = TensorDataset(test_x, test_y)
validation_loader = DataLoader(dataset=validation_data, batch_size=15, shuffle=True)

In [70]:
class SimpleNN(nn.Module):
    def __init__(self):
        super(SimpleNN, self).__init__()
        # declaring some parameters
        self.inputSize = 156
        self.outputSize = 2
        self.hidden1Size = 500
#         self.hidden2Size = 200
#         self.hidden3Size = 20

        # layers of linear transformation
        self.hidden1 = nn.Linear(self.inputSize, self.hidden1Size)
#         self.hidden2 = nn.Linear(self.hidden1Size, self.hidden2Size)
#         self.hidden3 = nn.Linear(self.hidden2Size, self.hidden3Size)
        self.output = nn.Linear(self.hidden1Size, self.outputSize)


        # # sigmoid activation and softmax output
        # self.sigmoid = nn.Sigmoid()
        # self.softmax = nn.Softmax(dim=1)

    def forward(self, x):

        x = torch.sigmoid(self.hidden1(x))
#         x = torch.sigmoid(self.hidden2(x))
#         x = torch.sigmoid(self.hidden3(x))
        y = F.softmax(self.output(x), dim=1)

        return y
    
model = SimpleNN()
optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.5)
criterion = nn.CrossEntropyLoss()

def train(epoch, log_interval=20):
    # set model to training mode
    model.train()

    # loop over batch from the training set
    for batch_idx, (data, target) in enumerate(train_loader):
        # Zero gradient buffers
        optimizer.zero_grad()

        # pass data through the network
        output = model(data)

        # calculate loss
        loss = criterion(output, target)

        # Backpropagate
        loss.backward()

        # Update Weights
        optimizer.step()

    print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
        epoch, batch_idx * len(data), len(train_loader.dataset),
        100. * batch_idx / len(train_loader), loss.data.item()))

    return

def validate(loss_vector, accuracy_vector):
    model.eval()
    val_loss, correct = 0, 0
    for data, target in validation_loader:
        output = model(data)
        val_loss += criterion(output, target).data.item()
        pred = output.data.max(1)[1]
        correct += pred.eq(target.data).sum()

        val_loss /= len(validation_loader)
        loss_vector.append(val_loss)
        accuracy = 100. * correct.to(torch.float32) / len(validation_loader.dataset)
        accuracy_vector.append(accuracy)

    print('\nValidation set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        val_loss, correct, len(validation_loader.dataset), accuracy))

In [71]:
num_epochs = 50
lossv, accv = [], []

for epoch in range(1, num_epochs + 1):
    train(epoch)
    validate(lossv, accv)

RuntimeError: 1D target tensor expected, multi-target not supported

In [ ]:
print('the fully-connected NN has max accuracy', 2124/2457)

# Voting Classifier

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import VotingClassifier
from sklearn import svm

In [ ]:
clf = VotingClassifier([('logr', LogisticRegressionCV(10, max_iter=800)),
                            ('rfor', RandomForestClassifier(n_estimators = 50, 
                                                            random_state = 0, max_depth=30)),
                       ('svm', svm.SVC())])

clf.fit(X_train, y_train)

In [60]:
vot_predicted = clf.predict(X_test)
vot_accuracy = accuracy_score(y_test, vot_predicted)
vot_accuracy

NameError: name 'clf' is not defined

In [ ]:
print('The voting classifier has accuracy', vot_accuracy)

# SVM 

In [ ]:
model_svm = svm.SVC() # I chose RBF kernel here to project into infinite dimensions
model_svm.fit(X_train, y_train)

In [ ]:
svm_predicted = model_svm.predict(X_test)
svm_accuracy = accuracy_score(y_test, svm_predicted)
svm_accuracy

In [ ]:
print('The svm using Radial Basis Function as kernel has accuracy', svm_accuracy)

# Visualization

In [ ]:
from sklearn import decomposition
from sklearn.manifold import TSNE
import matplotlib.pyplot as plt

In [ ]:
def visualization(decomposition_data, labels_pred, path):
    print('plotting visualization...')
    x = []
    y = []
    for i in decomposition_data:
        x.append(i[0])
        y.append(i[1])
    plt.figure(figsize=(20, 20))
    plt.axes()
    plt.scatter(x, y, c=labels_pred, marker=".")
    plt.xticks(())
    plt.yticks(())
    plt.savefig(path)
    print('plot saved to ' + path + '\n')
    return

In [ ]:
tsne = TSNE(n_components=2)
decomposition_data_test = tsne.fit_transform(X_test)

In [ ]:
visualization(decomposition_data_test, y_test, 'fig_truelabel.png')
visualization(decomposition_data_test, RF_predicted, 'fig_rf_predict.png')
visualization(decomposition_data_test, LR_predicted, 'fig_lr_predict.png')